File: app/src/main/java/com/simplyillustrate/simplyillustrate/models/Question.java
package com.simplyillustrate.simplyillustrate.models;

import java.util.ArrayList;
import java.util.List;

import android.os.Parcel;
import android.os.Parcelable;
import android.os.Parcelable.Creator;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

public class Question implements Parcelable {

    @SerializedName("tags")
    @Expose
    private List<String> tags = new ArrayList<>();
    @SerializedName("_id")
    @Expose
    private String id;
    @SerializedName("user")
    @Expose
    private String user;
    @SerializedName("title")
    @Expose
    private String title;
    @SerializedName("difficulty")
    @Expose
    private String difficulty;
    @SerializedName("description")
    @Expose
    private String description;
    @SerializedName("created_at")
    @Expose
    private String createdAt;
    @SerializedName("__v")
    @Expose
    private Integer v;
    public final static Parcelable.Creator<Question> CREATOR = new Creator<Question>() {


        @SuppressWarnings({
                "unchecked"
        })
        public Question createFromParcel(Parcel in) {
            return new Question(in);
        }

        public Question[] newArray(int size) {
            return (new Question[size]);
        }

    };

    protected Question(Parcel in) {
        in.readList(this.tags, (java.lang.String.class.getClassLoader()));
        this.id = ((String) in.readValue((String.class.getClassLoader())));
        this.user = ((String) in.readValue((String.class.getClassLoader())));
        this.title = ((String) in.readValue((String.class.getClassLoader())));
        this.difficulty = ((String) in.readValue((String.class.getClassLoader())));
        this.description = ((String) in.readValue((String.class.getClassLoader())));
        this.createdAt = ((String) in.readValue((String.class.getClassLoader())));
        this.v = ((Integer) in.readValue((Integer.class.getClassLoader())));
    }

    public Question() {
    }

    public List<String> getTags() {
        return tags;
    }

    public void setTags(List<String> tags) {
        this.tags = tags;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getUser() {
        return user;
    }

    public void setUser(String user) {
        this.user = user;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDifficulty() {
        return difficulty;
    }

    public void setDifficulty(String difficulty) {
        this.difficulty = difficulty;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(String createdAt) {
        this.createdAt = createdAt;
    }

    public Integer getV() {
        return v;
    }

    public void setV(Integer v) {
        this.v = v;
    }

    @Override
    public String toString() {
        return new StringBuilder().append("tags = "+tags).append("id = "+ id)
                .append("user = "+ user)
                .append("title = "+ title)
                .append("difficulty = "+ difficulty)
                .append("description = "+ description)
                .append("createdAt = "+ createdAt)
                .append("v = "+ v).toString();
    }

    public void writeToParcel(Parcel dest, int flags) {
        dest.writeList(tags);
        dest.writeValue(id);
        dest.writeValue(user);
        dest.writeValue(title);
        dest.writeValue(difficulty);
        dest.writeValue(description);
        dest.writeValue(createdAt);
        dest.writeValue(v);
    }

    public int describeContents() {
        return 0;
    }

}

File: app/src/main/java/com/simplyillustrate/simplyillustrate/entity/Question.kt
package com.simplyillustrate.simplyillustrate.entity

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import com.simplyillustrate.simplyillustrate.SimplyIllustrate
import java.util.*
import kotlin.collections.ArrayList

class Question() : Parcelable {

    @SerializedName("_id")
    var id: String? = null

    @SerializedName("created_by")
    var createdBy: User? = null

    @SerializedName("title")
    var title: String? = null

    @SerializedName("tags")
    var tags: ArrayList<Tag> = ArrayList()

    @SerializedName("difficulty")
    var difficulty: String? = null

    @SerializedName("description")
    var description: String? = null

    @SerializedName("timestamp")
    var timestamp: Date? = null

    @SerializedName("__v")
    var v: Int? = null

    fun getTagNames(): List<String> {
        return tags.map { tag ->
            tag.name ?: "no_name"
        }
    }

    constructor(parcel: Parcel) : this() {
        id = parcel.readString()
        createdBy = parcel.readParcelable(User::class.java.classLoader)
        title = parcel.readString()
        difficulty = parcel.readString()
        description = parcel.readString()
        timestamp = parcel.readSerializable() as? Date
        v = parcel.readValue(Int::class.java.classLoader) as? Int
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(id)
        parcel.writeParcelable(createdBy, flags)
        parcel.writeString(title)
        parcel.writeString(difficulty)
        parcel.writeString(description)
        parcel.writeSerializable(timestamp)
        parcel.writeValue(v)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Question> {
        override fun createFromParcel(parcel: Parcel): Question {
            return Question(parcel)
        }

        override fun newArray(size: Int): Array<Question?> {
            return arrayOfNulls(size)
        }
    }
}